static/js/containers/pages/checkout/CartPage.js

```diff
@@ -2,34 +2,140 @@
+import {
+  ALERT_TYPE_DANGER,
+  ALERT_TYPE_SUCCESS,
+  CART_DISPLAY_PAGE_TITLE
+} from "../../../constants"
 import { compose } from "redux"
 import { connect } from "react-redux"
-import { connectRequest } from "redux-query"
-import type { BasketItem } from "../../../flow/cartTypes"
+import { connectRequest, mutateAsync } from "redux-query"
+import { createStructuredSelector } from "reselect"
+
+import type { BasketItem, Discount } from "../../../flow/cartTypes"
 
 import Loader from "../../../components/Loader"
 
-import { createStructuredSelector } from "reselect"
 import {
   cartQuery,
   cartSelector,
-  totalPriceSelector
+  totalPriceSelector,
+  discountedPriceSelector,
+  discountSelector,
+  applyDiscountCodeMutation,
+  clearDiscountCodeMutation
 } from "../../../lib/queries/cart"
 
 import type { RouterHistory } from "react-router"
 import moment from "moment"
-import { formatPrettyDateTimeAmPmTz, parseDateString } from "../../../lib/util"
+import {
+  formatPrettyDateTimeAmPmTz,
+  parseDateString,
+  isSuccessResponse,
+  formatLocalePrice
+} from "../../../lib/util"
 import { Button } from "reactstrap"
+import { addUserNotification } from "../../../actions"
+
+import ApplyCouponForm from "../../../components/forms/ApplyCouponForm"
 
 type Props = {
   history: RouterHistory,
   cartItems: Array<BasketItem>,
   totalPrice: number,
-  isLoading: boolean
+  discountedPrice: number,
+  discounts: Array<Discount>,
+  isLoading: boolean,
+  applyDiscountCode: (code: string) => Promise<any>,
+  clearDiscountCode: () => Promise<any>,
+  addUserNotification: Function,
+  forceRequest: Function
+}
+
+type CartState = {
+  discountCode: string,
+  discountCodeIsBad: boolean
 }
 
-export class CartPage extends React.Component<Props> {
+export class CartPage extends React.Component<Props, CartState> {
+  state = {
+    discountCode:      "",
+    discountCodeIsBad: false
+  }
+
+  updateCode(event: Object) {
+    this.setState({
+      discountCode:      event.target.value,
+      discountCodeIsBad: false
+    })
+  }
+
+  async addDiscount(ev: Object) {
+    const subbedCode = ev.couponCode
+    const { applyDiscountCode, forceRequest, addUserNotification } = this.props
+
+    this.setState({ discountCode: subbedCode, discountCodeIsBad: false })
+
+    if (String(subbedCode).trim().length === 0) {
+      return
+    }
+
+    let userMessage, messageType
+
+    const resp = await applyDiscountCode(this.state.discountCode)
+    if (isSuccessResponse(resp)) {
+      messageType = ALERT_TYPE_SUCCESS
+      userMessage = "Discount code added."
+
+      forceRequest()
+    } else {
+      messageType = ALERT_TYPE_DANGER
+      userMessage = `Discount code ${this.state.discountCode} is invalid.`
+
+      this.setState({ discountCodeIsBad: true })
+    }
+
+    addUserNotification({
+      "add-discount-notification": {
+        type:  messageType,
+        props: {
+          text: userMessage
+        }
+      }
+    })
```

Comment:
I think it's fine to just drop this, it shifts the page layout, which is a bit distracting and there's already an inline error message.